Implement type union/intersection

rossberg opened the pull request:
Fixes #1841.

crusso commented on src/mo_def/arrange.ml:
```diff
@@ -178,6 +178,8 @@ and typ t = match t.it with
   | FuncT (s, tbs, at, rt) -> "FuncT" $$ [func_sort s] @ List.map typ_bind tbs @ [ typ at; typ rt]
   | AsyncT (t1, t2) -> "AsyncT" $$ [typ t1; typ t2]
+  | AndT (t1, t2) -> "AndT" $$ [typ t1; typ t2]
+  | OrT (t1, t2) -> "AndT" $$ [typ t1; typ t2]
```
```suggestion
  | OrT (t1, t2) -> "OrT" $$ [typ t1; typ t2]
```

crusso reviewed (commented):
This looks lovely, but haven't had a chance to review properly yet. Maybe tomorrow.

crusso commented on doc/modules/language-guide/pages/language-manual.adoc:
```diff
@@ -902,6 +904,22 @@ No value has type `None`.
+[[or-type]]
+=== Union type
+
+The type expression `<typ1> or <typ2>` denotes the syntactic _union_ between its two type operands, that is, the smallest type that is a supertype of both. If both types are incompatible, the intersection is `Any`.
```
```suggestion
The type expression `<typ1> or <typ2>` denotes the syntactic _union_ between its two type operands, that is, the smallest type that is a supertype of both. If both types are incompatible, the union is `Any`.
```

crusso commented on src/mo_frontend/typing.ml:
```diff
@@ -476,6 +476,28 @@ and check_typ' env typ : T.typ =
+  | AndT (typ1, typ2) ->
+    let t1 = check_typ env typ1 in
+    let t2 = check_typ env typ2 in
+    let t = T.glb t1 t2 in
```
Can't this lead to non-termination if t1 or t2 is sibling recursive type that isn't known to be contractive yet?

What does:
 "type A = A and/or A"
 produce?

crusso replied:
(Type elaboration used to be syntactic with some checks, now it ain't.)

rossberg replied:
Good question. Intersection and union are not productive themselves, so this produces:
```
type error [M0157], block contains non-productive definition A
```
I think this works fine because we do that check before elaborating the types. I added a couple of simple tests.
> (Type elaboration used to be syntactic with some checks, now it ain't.)
True, is that a problem?

rossberg replied:
Ah, okay, this example happens to work, because in the prepass a named type is simply represented as its abstract name and union/intersection shortcuts the reflexive case.

Here is a problematic case:
```
type T = U and V; type U = {}; type V = {};
```
This crashes with an assertion failure, because glb computation needs to expand the type id and runs into type `Pre` in the prepass. Hm...

crusso replied:
Bingo, thanks for persevering.

rossberg replied:
Yeah, I thought about making them more primitive, but then I'd rather bite and make them proper intersection and union types.

Another option I considered is to check that the operands aren't Pre, and if they are, give an "cannot compute intersection on forward types" error, similar to what we do for type inference in similar cases.

rossberg replied:
Okay, I implemented the latter. Extending our type checker's "incompleteness" that way seems okay given that we're already incomplete in similar ways. And it's conservative, so we can do smarter later if this ever turns out to be relevant in practice.

rossberg commented:
Thanks! Changed error message to "recursive or forward references", since the error doesn't actually depend on recursion (cf. my counter example). Also added a couple of recursive examples to test.

FWIW, order does matter here, even in the recursive case, because type definitions are still computed in order. This is rejected:
```
type B = A and {}; type A = {a : B};
```
while this is accepted:
```
type A = {a : B}; type B = A and {};
```
I know, it's a bit unsatisfactory... Hope we get to turn these into proper union/intersection types eventually.

jzxchiang1 commented:
Can this be added to the official SDK [documentation](https://smartcontracts.org/docs/language-guide/overview.html)? It's a really useful feature.
